### PR TITLE
Reduction to QPs

### PR DESCRIPTION
**Pull request by roversch**
This is the work I've done so far on the reductions, including reduction to QPs. The rewritten cone interfaces I pushed earlier, to be able to collaborate with @enzobusseti .

My code passes the tests `test_linear_cone.py` and `test_qp.py`, and there should be no `flake8` errors.

Let me know when you have comments!

### CODE REVIEW

File: cvxpy/atoms/atom.py [hunk 1]
```diff
@@ -70,6 +70,12 @@ def sign_from_args(self):
         """
         return NotImplemented
 
+    @property
```
Comment (SteveDiamond) [inline cvxpy/atoms/atom.py]: What is this?

File: cvxpy/atoms/atom.py [hunk 2]
```diff
@@ -128,7 +134,7 @@ def is_convex(self):
         else:
             return False
 
-    @clru_cache(maxsize=100)
+    # @clru_cache(maxsize=100)
```
Comment (SteveDiamond) [inline cvxpy/atoms/atom.py]: Why did you comment these out?

File: cvxpy/atoms/quad_form.py [hunk 3]
```diff
@@ -18,20 +18,138 @@
...
+        return self.P * self.x
+
+    def graph_implementation(self):
+        return NotImplemented
+
+    def shape_from_args(self):
+        return (1, 1)
+
+
+class SymbolicQuadForm(Atom):
```
Comment (SteveDiamond) [inline cvxpy/atoms/quad_form.py]: What is ``expr``?

File: cvxpy/reductions/matrix_stuffing.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+"""
```
Comment (SteveDiamond) [inline cvxpy/reductions/matrix_stuffing.py]: What is this file supposed to be? The name is unclear.

File: cvxpy/reductions/qp2quad_form/qp_matrix_stuffing.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            all([not v.domain for v in prob.variables()])  # no implicit variable domains
+            # (TODO: domains are not implemented yet)
+        )
+
+    def apply(self, problem):
+        """Returns a new problem and data for inverting the new solution."""
+        qp, inverse_data_stack = Qp2SymbolicQp().apply(problem)
+
+        if not self.accepts(qp):
+            raise ValueError("This QP can not be stuffed")
```
Comment (SteveDiamond) [inline cvxpy/reductions/qp2quad_form/qp_matrix_stuffing.py]: Haha I like this error.

File: cvxpy/reductions/qp2quad_form/replace_quad_forms.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+from cvxpy.atoms.quad_form import SymbolicQuadForm, QuadForm
+from cvxpy.expressions.variables import Variable
+from cvxpy.lin_ops.lin_op import NO_OP, LinOp
+from cvxpy.problems.objective import Maximize, Minimize
+from cvxpy.problems.problem import Problem
+from cvxpy.reductions.reduction import Reduction
+
+
+class ReplaceQuadForms(Reduction):
+    """Replaces symbolic quadratic forms in the objective by ordinary variables"""
```
Comment (SteveDiamond) [inline cvxpy/reductions/qp2quad_form/replace_quad_forms.py]: What does this mean?

**Comment by SteveDiamond** [comment]
Ok I made some comments. Code looks good overall. Mainly a lack of commenting/my unclearness on what's happening.

File: cvxpy/atoms/atom.py (hunk 1 above)
Comment (roversch) [inline cvxpy/atoms/atom.py]: Just a shorthand for `args[0]`, which is not so nice for e.g. `Minimize`, which has only one argument, namely the expression. I'll move it to Canonical.

File: cvxpy/atoms/atom.py (hunk 2 above)
Comment (roversch) [inline cvxpy/atoms/atom.py]: I implemented QP canonicalization with mutable trees. I will put this back and make a new copy of the tree instead.

File: cvxpy/atoms/quad_form.py (hunk 3 above)
Comment (roversch) [inline cvxpy/atoms/quad_form.py]: The QP canonicalization uses this, but I intend to rewrite this. Baiscally, it needs to keep track of the original expression (e.g. `quad_over_lin`), and then substitute a `SymbolicQuadForm` in this original expression.

I'm sure there is a better way of doing this.

File: cvxpy/reductions/matrix_stuffing.py (hunk 4 above)
Comment (roversch) [inline cvxpy/reductions/matrix_stuffing.py]: It's the super class for `QpMatrixStuffing` and `ConeMatrixStuffing`. I should probably make this abstract.

File: cvxpy/reductions/qp2quad_form/replace_quad_forms.py (hunk 6 above)
Comment (roversch) [inline cvxpy/reductions/qp2quad_form/replace_quad_forms.py]: CVXcanon only takes `Variable`s, so I need to replace the `SymbolicQuadForm`s with ordinary variables. These variables are just placeholders.

File: cvxpy/atoms/atom.py (hunk 1 above)
Comment (roversch) [inline cvxpy/atoms/atom.py]: done

File: cvxpy/reductions/matrix_stuffing.py (hunk 4 above)
Comment (roversch) [inline cvxpy/reductions/matrix_stuffing.py]: done

**Pull request merged**